flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -142,14 +147,33 @@ public NetworkBufferPool(
 
 	@Nullable
 	public MemorySegment requestMemorySegment() {
-		return availableMemorySegments.poll();
+		synchronized (availabilityHelper) {
+			final MemorySegment segment = availableMemorySegments.poll();
+			if (availableMemorySegments.isEmpty() && segment != null) {
+				availabilityHelper.resetUnavailable();
+			}
+			return segment;
+		}
 	}
 
 	public void recycle(MemorySegment segment) {
 		// Adds the segment back to the queue, which does not immediately free the memory
 		// however, since this happens when references to the global pool are also released,
 		// making the availableMemorySegments queue and its contained object reclaimable
-		availableMemorySegments.add(checkNotNull(segment));
+		checkNotNull(segment);
+		CompletableFuture<?> toNotify = null;
```

Comment:
If possible we need to consider deduplicating codes with below `recycleMemorySegments(Collection<MemorySegment> segments, int size)`. There is only `availableMemorySegments.add(segment);` different.